fud2/fud-core/src/script/plugin.rs

```diff
@@ -72,6 +86,206 @@ impl ScriptContext {
+        output_names: rhai::Array,
+        outputs: rhai::Array,
+    ) -> RhaiResult<()> {
+        let inputs = inputs
+            .into_iter()
+            .map(|i| match i.clone().try_cast::<StateRef>() {
+                Some(state) => Ok(state),
+                None => Err(RhaiSystemError::state_ref(i).with_pos(pos).into()),
+            })
+            .collect::<RhaiResult<Vec<_>>>()?;
+        let input_names = input_names
+            .into_iter()
+            .map(|i| match i.clone().try_cast::<String>() {
+                Some(state) => Ok(state),
+                None => Err(RhaiSystemError::state_ref(i).with_pos(pos).into()),
+            })
+            .collect::<RhaiResult<Vec<_>>>()?;
+        let outputs = outputs
+            .into_iter()
+            .map(|i| match i.clone().try_cast::<StateRef>() {
+                Some(state) => Ok(state),
+                None => Err(RhaiSystemError::state_ref(i).with_pos(pos).into()),
+            })
+            .collect::<RhaiResult<Vec<_>>>()?;
+        let output_names = output_names
+            .into_iter()
+            .map(|i| match i.clone().try_cast::<String>() {
+                Some(state) => Ok(state),
+                None => Err(RhaiSystemError::state_ref(i).with_pos(pos).into()),
+            })
+            .collect::<RhaiResult<Vec<_>>>()?;
+        let input_tuples = input_names.into_iter().zip(inputs).collect();
+        let output_tuples = output_names.into_iter().zip(outputs).collect();
+
+        let mut cur_op = self.cur_op.borrow_mut();
+        match *cur_op {
+            None => {
+                *cur_op = Some((
+                    name.to_string(),
+                    input_tuples,
+                    output_tuples,
+                    vec![],
+                ));
+                Ok(())
+            }
+            Some((ref old_name, _, _, _)) => {
+                Err(RhaiSystemError::began_op(old_name, name)
+                    .with_pos(pos)
+                    .into())
+            }
+        }
+    }
+
+    /// Adds a shell command to the `cur_op`.Returns and error if `begin_op` has not been called
+    /// before this `end_op` and after any previous `end_op`
+    fn add_shell(&self, pos: Position, cmd: String) -> RhaiResult<()> {
+        let mut cur_op = self.cur_op.borrow_mut();
+        match *cur_op {
+            Some(ref mut op_sig) => {
+                op_sig.3.push(cmd);
+                Ok(())
+            }
+            None => Err(RhaiSystemError::no_op().with_pos(pos).into()),
+        }
+    }
+
+    /// If `#ident` is a substring of `cmd` and `("ident", "value")` is in `mapping`, returns a
+    /// string identical to `cmd` except for all instances `#ident` replaced with `value`.
+    fn substitute_shell_text(cmd: &str, mapping: &[(&str, &str)]) -> String {
+        let mut cur = cmd.to_string();
+        for (i, v) in mapping {
+            let substr = format!("#{}", i);
+            cur = cur.replace(&substr, v);
```

Comment:
This is one place where we probably need to worry about shell escaping. (But not in this PR, IMO.)

Reply:
This is now a non-issue because I'm using Ninja variables now instead of some special escape string.